src/test/java/org/mockitousage/serialization/DeepStubsSerializableTest.java

```diff
@@ -45,18 +49,22 @@ public void should_serialize_and_deserialize_parameterized_class_mocked_with_dee
         assertThat(deserialized_deep_stub.iterator().next().add("yes")).isEqualTo(true);
     }
 
-    @Test(expected = NullPointerException.class)
+    @Test
     public void should_discard_generics_metadata_when_serialized_then_disabling_deep_stubs_with_generics() throws Exception {
         // given
         ListContainer deep_stubbed = mock(ListContainer.class, withSettings().defaultAnswer(RETURNS_DEEP_STUBS).serializable());
         when(deep_stubbed.iterator().hasNext()).thenReturn(true);
 
         ListContainer deserialized_deep_stub = serializeAndBack(deep_stubbed);
 
-        // when stubbing on a deserialized mock
-        when(deserialized_deep_stub.iterator().next().get(42)).thenReturn("no");
-
-        // then revert to the default RETURNS_DEEP_STUBS and the code will raise a ClassCastException
+        try {
+            // when stubbing on a deserialized mock
+            // then revert to the default RETURNS_DEEP_STUBS and the code will raise a ClassCastException
+            when(deserialized_deep_stub.iterator().next().get(42)).thenReturn("no");
+            fail("Expected an exception to be thrown as deep stubs and serialization does not play well together");
```

Comment:
This is a regression. Can we not throw a CCE here? If not, we are re-introducing #357.

Reply:
Nope we're not this change actually properly fix the issue

Reply:
I'm working on the next commit for a complete fix. Anyway, with the build matrix, the unit tests do fail on JDK > 8.

Reply:
@TimvdLippe Plus this test is about documenting the issue about deserialized mock.
Do you have a lot at google ?

Reply:
I don't recall the actual numbers, but if I would have to guess it was probably between 50 and 200.

Note that this test is slightly different than the pattern we saw the regression on. The tests were doing something along the lines of `when(subject.performBla().performOtherThing().getValue()).thenReturn(something)`. The `getValue()` in this case returned `T`. Without the `return null`, it would throw a CCE on the `.getValue()` invocation. With the `return null`, it would safely pass that and properly set up the stub.

This test case goes one step further: `when(subject.performBla().performOtherThing().getValue().otherInvocation()).thenReturn(something)`. The `otherInvocation` has never worked. Previously, it would already throw a CCE on the `getValue()`. Now, it throws a NPE on the `otherInvocation()`. E.g. we didn't regress, we are just throwing an exception in a different way.

So the former I don't want to regress on. I am fine in the latter situation to make changes. (E.g. this test in particular). If we are speaking in terms of this test, we need `when(deserialized_deep_stub.iterator().next()).thenReturn(object)` to succeed.

Reply:
Oh and if your fix does not regress in the first situation, that is more appropriate than my `return null`. So I am all for that!

Reply:
I don't know the codebase your dealing with, but the resolver is tad more intelligent now, now there's still some ground that is not covered. For that reason I decided to add back the guard, with an additional check.

So that way it should keep the old fix places in cases the generic metadata support fails to resolve generic types.

Reply:
Awesome, taking a look :smile: